lib/active_job/limiter/mixin.rb

```diff
@@ -29,6 +29,68 @@ def limit_queue(expiration:)
             ActiveJob::Limiter.clear_lock_before_perform(job)
           end
         end
+
+        def throttle_job(duration:, extract_resource_id:, metrics_hook: ->(_result, _job) {})
+          active_job_limiter_add_throttle_job_around_enqueue(
+            duration: duration,
+            extract_resource_id: extract_resource_id,
+            metrics_hook: metrics_hook
+          )
+          active_job_limiter_add_throttle_job_around_perform(
+            duration: duration,
+            extract_resource_id: extract_resource_id,
+            metrics_hook: metrics_hook
+          )
+        end
+
+        private
+
+        # Prefixing private methods with active_job_limiter to avoid conflicts with user code
+
+        def active_job_limiter_add_throttle_job_around_enqueue(duration:, extract_resource_id:, metrics_hook:)
+          around_enqueue do |job, block|
+            resource_id = extract_resource_id.call(job)
+
+            if job.instance_variable_get(:@bypass_active_job_limiter_enqueue_locks)
+              block.call
+            elsif ActiveJob::Limiter.acquire_lock_for_job_resource('enqueue', duration, job, resource_id)
+              block.call
+              metrics_hook.call('enqueue.enqueued', job)
+            else
+              job.job_id = nil
+              metrics_hook.call('enqueue.dropped', job)
+            end
+          end
+        end
+
+        def active_job_limiter_add_throttle_job_around_perform(duration:, extract_resource_id:, metrics_hook:)
+          around_perform do |job, block|
+            resource_id = extract_resource_id.call(job)
+
+            if ActiveJob::Limiter.acquire_lock_for_job_resource('perform', duration, job, resource_id)
+              ActiveJob::Limiter.release_lock_for_job_resource('enqueue', job, resource_id)
+              block.call
+              metrics_hook.call('perform.performed', job)
+            elsif ActiveJob::Limiter.acquire_lock_for_job_resource('reschedule', duration, job, resource_id)
```

Comment:
when is this lock released?

Reply:
It is automatically released by redis after the specified duration (it adds an expire to the key).